crypto/noise.c

```diff
@@ -0,0 +1,300 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES
+ * OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "noise.h"
+#include <string.h>
+
+#include "aes/aesgcm.h"
+#include "ed25519-donna/ed25519.h"
+#include "hmac.h"
+#include "rand.h"
+#include "sha2.h"
+
+_Static_assert(KEY_SIZE == SHA256_DIGEST_LENGTH,
+               "KEY_SIZE must be equal to SHA256_DIGEST_LENGTH");
+
+static uint8_t protocol_name[32] = {'N', 'o', 'i', 's', 'e', '_',  'K',  'K',
+                                    '1', '_', '2', '5', '5', '1',  '9',  '_',
+                                    'A', 'E', 'S', 'G', 'C', 'M',  '_',  'S',
+                                    'H', 'A', '2', '5', '6', 0x00, 0x00, 0x00};
+
+static bool encrypt(const uint8_t key[KEY_SIZE],
+                    const uint8_t nonce[NONCE_SIZE],
+                    const uint8_t *associated_data,
+                    size_t associated_data_length, const uint8_t *plaintext,
+                    size_t plaintext_length, uint8_t *ciphertext) {
+  // ciphertext = AES-GCM-Encrypt(key, nonce, associated_data, plaintext)
+  const size_t ciphertext_length = plaintext_length + TAG_SIZE;
+
+  uint8_t buffer[plaintext_length + TAG_SIZE];
+  memcpy(buffer, plaintext, plaintext_length);
+
+  gcm_ctx ctx = {0};
+  if (gcm_init_and_key(key, KEY_SIZE, &ctx) != RETURN_GOOD) {
+    return false;
+  }
+  if (gcm_encrypt_message(nonce, NONCE_SIZE, associated_data,
+                          associated_data_length, buffer, plaintext_length,
+                          buffer + plaintext_length, TAG_SIZE,
+                          &ctx) != RETURN_GOOD) {
+    return false;
+  }
+
+  memcpy(ciphertext, buffer, ciphertext_length);
+  return true;
+}
+
+static bool decrypt(const uint8_t key[KEY_SIZE],
+                    const uint8_t nonce[NONCE_SIZE],
+                    const uint8_t *associated_data,
+                    size_t associated_data_length, const uint8_t *ciphertext,
+                    size_t ciphertext_length, uint8_t *plaintext) {
+  // plaintext = AES-GCM-Decrypt(key, nonce, associated_data, ciphertext)
+  if (ciphertext_length < TAG_SIZE) {
+    return false;
+  }
+  const size_t plaintext_length = ciphertext_length - TAG_SIZE;
+
+  uint8_t buffer[ciphertext_length];
+  memcpy(buffer, ciphertext, ciphertext_length);
+
+  gcm_ctx ctx = {0};
+  if (gcm_init_and_key(key, KEY_SIZE, &ctx) != RETURN_GOOD) {
+    return false;
+  }
+  if (gcm_decrypt_message(nonce, NONCE_SIZE, associated_data,
+                          associated_data_length, buffer, plaintext_length,
+                          buffer + plaintext_length, TAG_SIZE,
+                          &ctx) != RETURN_GOOD) {
+    return false;
+  }
+
+  memcpy(plaintext, buffer, plaintext_length);
+  return true;
```

Comment:
memzero `buffer` before return or even better, just use directly the `plaintext` buffer that is already allocated.